enclave-runtime/src/lib.rs

```diff
@@ -486,188 +476,31 @@ pub unsafe extern "C" fn sync_parentchain(
-					error!("Error performing shield funds. Error: {:?}", e);
-				} else {
-					// Cache successfully executed shielding call.
-					executed_shielding_calls.push(hash_of(xt))
-				}
-			}
-		};
-
-		// Found CallWorker extrinsic in block.
-		if let Ok(xt) =
-			UncheckedExtrinsicV4::<CallWorkerFn>::decode(&mut xt_opaque.encode().as_slice())
-		{
-			if xt.function.0 == [TEEREX_MODULE, CALL_WORKER] {
-				if let Ok((decrypted_trusted_call, shard)) = decrypt_unchecked_extrinsic(xt) {
-					if let Err(e) = stf_executor.execute_trusted_call::<PB>(
-						&mut opaque_calls,
-						&decrypted_trusted_call,
-						block.header(),
-						&shard,
-						StatePostProcessing::Prune, // we only want to store the state diff for direct stuff.
-					) {
-						error!("Error executing trusted call: Error: {:?}", e);
-					}
-				}
-			}
-		}
-	}
-	Ok((opaque_calls, executed_shielding_calls))
-}
-
-fn hash_of<T: Encode>(xt: T) -> H256 {
-	blake2_256(&xt.encode()).into()
-}
-
-fn handle_shield_funds_xt<StfExecutor>(
-	xt: &UncheckedExtrinsicV4<ShieldFundsFn>,
-	stf_executor: &StfExecutor,
-) -> Result<()>
-where
-	StfExecutor: StfUpdateState + StfExecuteTrustedCall + StfExecuteShieldFunds,
-{
-	let (call, account_encrypted, amount, shard) = &xt.function;
-	info!("Found ShieldFunds extrinsic in block: \nCall: {:?} \nAccount Encrypted {:?} \nAmount: {} \nShard: {}",
-        call, account_encrypted, amount, shard.encode().to_base58(),
-    );
-
-	debug!("decrypt the call");
-	let account_vec = Rsa3072Seal::unseal().map(|key| key.decrypt(account_encrypted))??;
-
-	let account = AccountId::decode(&mut account_vec.as_slice())
-		.sgx_error_with_log("[ShieldFunds] Could not decode account")?;
-
-	stf_executor.execute_shield_funds(account, *amount, shard)?;
-	Ok(())
-}
+		Arc::new(ExtrinsicsFactory::new(genesis_hash, signer, GLOBAL_NONCE_CACHE.clone()));
+	let indirect_calls_executor =
+		Arc::new(IndirectCallsExecutor::new(shielding_key, stf_executor.clone()));
+	let parentchain_block_importer = ParentchainBlockImporter::new(
+		validator_access,
+		Arc::new(OcallApi),
+		stf_executor,
+		extrinsics_factory,
+		indirect_calls_executor,
+	);
 
-fn decrypt_unchecked_extrinsic(
-	xt: UncheckedExtrinsicV4<CallWorkerFn>,
-) -> Result<(TrustedCallSigned, ShardIdentifier)> {
-	let (call, request) = xt.function;
-	let (shard, cyphertext) = (request.shard, request.cyphertext);
-	debug!("Found CallWorker extrinsic in block: \nCall: {:?} \nRequest: \nshard: {}\ncyphertext: {:?}",
-        call,
-        shard.encode().to_base58(),
-        cyphertext
-    );
-
-	debug!("decrypt the call");
-	//let request_vec = Rsa3072KeyPair::decrypt(&cyphertext)?;
-	let request_vec = Rsa3072Seal::unseal().map(|key| key.decrypt(&cyphertext))??;
-
-	Ok(TrustedCallSigned::decode(&mut request_vec.as_slice()).map(|call| (call, shard))?)
+	parentchain_block_importer
+		.import_parentchain_blocks(blocks_to_sync)
+		.map_err(|e| e.into())
```

Comment:
Removed a whole lot of code. What remains is the creation of a `ParentchainBlockImporter` on which we call `import_parentchain_blocks`.